### DISCUSSION TITLE
Palm Springs Hydrojet #54144 4-wire E13 Error

### DISCUSSION THREAD

**Discussion by OnurbHartmann** [visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA#150] [Q&A]
Hello all,
It took a while for me to set things up and go for the first testing.
But no luck yet.
After connecting and powering on, the system shows up with the following behavior
- for about 5 sec the DSP shows a temperature (does not match with real temp in the spa)
- then E13 blinking and beep
- the blue LED on ESP is not constantly on, show weird behavior, flashing or on for a few seconds, however repeating

On the bench the ESP runs with blue LED on and http access.
Searching in discussions I came across the #91 where @Bishbosh02 mentioned a different pinout.

Should I just give it a try and use that wiring?
Any other thoughts?

Here is the CIO connector. No luck on reading the text beneath.

![IMG_1032](https://user-images.githubusercontent.com/32487897/142937032-de611a53-464b-49e7-b4df-2ffd59800c7b.jpg)

**Comment by Inazone** [comment]
Your issue sounds similar to mine. See issue #144. I'm wondering if the 8266 is playing up. Currently got another one on order, so will try that when it arrives.

**Comment by Bishbosh02** [comment]
Looking at the colours I would say the wiring is correct for your model BK=0V,RD = 5V, YL and GN is Tx and Rx, I dont have this model so cant check

**Comment by OnurbHartmann** [comment]
Thanks for the answers. 
I analyzed further. Just connecting the CIO to the board and powering the ESP runs and is accessible via web. 
Then connecting the DSP - the ESP goes into the rebooting mode, described above, and DSP shows E13 error.
Confused on that behavior. I checked the soldering multiple times, no issues found.
Then removed RX/TX from the wiring to DSP, just having the GND and 5V connected to DSP, same behavior. 
Checking the voltages still on 4.9V on Vin on ESP, and 3.3V on respective pins.
No clue where to go from here ...

**Reply by visualapproach** [reply]
Thinking out loud here:
Sounds like a power issue. I think it would be possible to check by disconnecting the 5V from the source (CIO). Then power the ESP with a powerbank to the USB. You do not want to have both 5V from CIO and USB because that might release the magic smoke that hides inside some of the components. Keep RX, TX and GND connected to both CIO and DSP. If it works now, we can be fairly confident that there is a power issue. It may help to put a capacitor over 5V and GND close to the ESP. Value - do not know but try a small one like a couple of uF to a 100 uF. Check polarity + to + if needed.

**Reply by OnurbHartmann** [reply]
Thanks, I tried and the USB power bank solved the power issue. However, E13 error occurs after about 10sec.
Reading through I will try the dev branch and an additional capacitor.

**Reply by visualapproach** [reply]
Nice, I think I will add a delay before the wifi starts. I guess the pump is power hungry at start-up, and so is the wifi. Worth a try at least.

**Reply by Bishbosh02** [reply]
just a note the E13 error is bad earth on the main module it can be a bit fussy if you are provided power via extension leads ect there a quite a few YouTube vids on this error.

**Reply by skyluker** [reply]
> Thanks, I tried and the USB power bank solved the power issue. However, E13 error occurs after about 10sec. Reading through I will try the dev branch and an additional capacitor.

@Bishbosh02 Hi, did you solve this issue? If so, could you provide a solution?

**Comment by DeanozUK** [comment]
First things first @visualapproach, Well done on this excellent project, I've been looking for this type thing for ages. Looking through your code amases me. the fact you are making this adapable to different version/models is a hard thing. Pat on the back m8.

Hi @OnurbHartmann, Did you ever get round to finding out what is causing this?

Ive a Hawaii Hydrojet pro (2019) #54138 and am getting same issue, tried 2 different ESP's, Converters, pcb and leads. Its exactly as you describe.

Background - been trying this for days. Wifi AP also moved near So not signal related and confirmed. will stay connected allday long with DSP disconnected. Also have 47uF cap on PCB 5v line, tried the 100uF too, but think this is for random brown outs some people report not this issue.

Plug in module, powers up DSP, 10sec later E13. Blue light is goes out but looks like its trying to flash (bootup)
Unplug DSP and blue light on and flashes. (CIO Active - Dev branch) but no control (even with 'take control' on) looks good so far.
Plug in DSP and 10secs later E13 and light goes out.

Tried disconnecting 5v from CIO to PCB and powered PCB, ESP and DSP from powerbank. Get same results however the ESP is either rebooting as wifi keeps going down on ESP so web page might come up briefly then timesout, is ESP crashing or Interference causing disconnect? I have no idea now and am loosing my mind.

Im going to see if i can workout how to flash this test passthrough code to confirm hardware.

**Reply by skyluker** [reply]
Did you manage to make this work?

**Reply by OnurbHartmann** [reply]
Hello @DeanozUK, after the inital setup i got frustrated and switched to work on my smarthome stuff. As this is all up and running I tested yesterday with latest code, still the same iussue with power. I doupt if the wiring is correct. Any further results or news from your investigation?

**Reply by cobaltfish** [reply]
I'm just getting going with my Palm Springs Hydrojet #54144. 
I've followed a lot of the other troubleshooting, and shortened my jst fly leads from 120mm to about 15mm, but was still getting e13.  I tried adding 100uF cap between gnd & 5v , and 10k resistor on the 3v, and added some ferrite beads on the cio and disp cables.  
I have a spare pump with failed heaters, and was bench testing against that (and a de-scaler unit with fresh water to allow pump to run), and got it stable on the dev branch.  

When I went to move it to the real item, it just kept dropping out.  it looked like as I plugged the display in, the NodeMcu rebooted.  I did note reboot reason was showing as hardware watch dog. I tried a number of ways of running the power without much luck, usb power banks, a 5v supply etc, but same result.  
I'm thinking one of 3 things could have been at play:
1. Slightly further from wifi AP, so drawing more power / generating more emi noise upsetting the serial coms. 
2. 5v supply on bench test pump a little more reliable / stable than the one I want to use it on, so things resetting
3. something upsetting the 3v supply / regulator, causing the txs0108e or the 8266 to reset, meaning it didn't relay data between cio and disp

I was about to give up, when I looked at [Issue 208](https://github.com/visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA/issues/208) again and saw the post from DeanzoUK with [super short fly leads ](https://github.com/visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA/issues/208#issuecomment-1029493163) and dug out a fresh board and tried again.  I used a bit of hotmelt to hold the plug/socket to the board to avoid too much flex on the soldered joints. Also, this time with no capacitors / resistors, but may try adding them later if it does play up.

I've also decided to not connect the Vin on the nodemcu, and have got a usb power supply patched into the power feeds off the main pump board ( a couple of crimped forked connectors onto the screw clamp / terminals) 

Only about an hour so far, but thats still many times better than the instant fail I was getting.

I've done a little bit of tweaking which I think should mean I can schedule "God mode" to allow for scheduled of peak heating, even if it has not been set.  I also need to remove the options for bubbles and ensure the jets show up in the web pages.

**Reply by visualapproach** [reply]
Thanks for the detailed description! I'm thinking the hardware watchdog reset is strange. Something is apparently locking up the cpu for too long but I don't know what. I think you should try lowering the websockets timeout to 2 s or something. There is a thread here somewhere about that.
/VA

**Reply by cobaltfish** [reply]
@visualapproach It was about 10pm, and I was feeling frustrated, so I didn't note down exactly what I saw, but am pretty certain it wasn't my imagination.
I did see that thread somewhere. I'm new to all of this, but if I've understood docs right, I might have tried to enable the aysnc web method, rather trying to  recompile libs.
In platformio.ini I added [me-no-dev/ESPAsyncTCP](https://github.com/me-no-dev/ESPAsyncTCP) into lib deps, and then added a build flag which I think should enable async: `build_flags =
    -DWEBSOCKETS_NETWORK_TYPE="NETWORK_ESP8266_ASYNC"  ` 
I'm not quite sure how to debug if it actually is using async or not! but it seems to load the web pages quicker

If I do get somewhere stable(ish) with E13's I'll get a branch pushed up

**Reply by visualapproach** [reply]
@cobaltfish lovely! Would be terrific if we could eliminate the problem. Unfortunately I can't give any advice on the ESPAsyncTCP because I have never used it. But I bet you need to change the ´´´#include <WebSocketsServer.h>´´´ in main.h to whatever library name that is.

**Reply by cobaltfish** [reply]
> @cobaltfish lovely! Would be terrific if we could eliminate the problem. Unfortunately I can't give any advice on the ESPAsyncTCP because I have never used it. But I bet you need to change the ´´´#include <WebSocketsServer.h>´´´ in main.h to whatever library name that is.

@visualapproach 
I saw on the [websockets readme that](https://github.com/Links2004/arduinoWebSockets#esp-async-tcp) that if the WEBSOCKETS_NETWORK_TYPE flag is set, it will then use the async lib.  I was surprised how simple it seemed. 
But that does seem to match on a quick scan of the codebase:
[Search WEBSOCKETS_NETWORK_TYPE](https://github.com/Links2004/arduinoWebSockets/search?q=WEBSOCKETS_NETWORK_TYPE)

I managed about 14 hours last night, then accidentally turned it off, when I meant to switch another socket on! But it successfully ran overnight, heating for a few hours during off peak electricity:

![image](https://user-images.githubusercontent.com/104502877/165704287-072658f1-4e46-4813-9a26-9b08b5476286.png)

**Comment by DeanozUK** [comment]
Hi Cobaltfish,Wow, that's a very detailed report of alot of things I tried and gave up as was going round in circles. Ive borrowed an oscilloscope and keep meaning to try and sniff the data in and out as I think it maybe slightly out of sync possibly. It's only a theory as tried so much. The reason for this is I noticed the absence of a crystal on both types of my  NodeMcus. Thus it's prob running software clock instead.....mayyyybe.Have you tried the other passthrough code Visual advised me. I got to work for several hours alone without touching then bam E13 no reason. Tried several times even with short leads and would e13 hours later.Let me know Ur findings, I'll have to dig the scope out and look now.DeanoOn 27 Apr 2022 13:44, cobaltfish ***@***.***> wrote:
I'm just getting going with my Palm Springs Hydrojet #54144.
I've followed a lot of the other troubleshooting, and shortened my jst fly leads from 120mm to about 15mm, but was still getting e13.  I tried adding 100uF cap between gnd & 5v , and 10k resistor on the 3v, and added some ferrite beads on the cio and disp cables.
I have a spare pump with failed heaters, and was bench testing against that (and a de-scaler unit with fresh water to allow pump to run), and got it stable on the dev branch.
When I went to move it to the real item, it just kept dropping out.  it looked like as I plugged the display in, the NodeMcu rebooted.  I did note reboot reason was showing as hardware watch dog. I tried a number of ways of running the power without much luck, usb power banks, a 5v supply etc, but same result.
I'm thinking one of 3 things could have been at play:
Slightly further from wifi AP, so drawing more power / generating more emi noise upsetting the serial coms.5v supply on bench test pump a little more reliable / stable than the one I want to use it on, so things resettingsomething upsetting the 3v supply / regulator, causing the txs0108e or the 8266 to reset, meaning it didn't relay data between cio and disp
I was about to give up, when I looked at Issue 208 again and saw the post from DeanzoUK with super short fly leads  and dug out a fresh board and tried again.  I used a bit of hotmelt to hold the plug/socket to the board to avoid too much flex on the soldered joints. Also, this time with no capacitors / resistors, but may try adding them later if it does play up.
I've also decided to not connect the Vin on the nodemcu, and have got a usb power supply patched into the power feeds off the main pump board ( a couple of crimped forked connectors onto the screw clamp / terminals)
Only about an hour so far, but thats still many times better than the instant fail I was getting.
I've done a little bit of tweaking which I think should mean I can schedule "God mode" to allow for scheduled of peak heating, even if it has not been set.  I also need to remove the options for bubbles and ensure the jets show up in the web pages.

**Reply by cobaltfish** [reply]
I've not got access to a logic analyser / oscilloscope, but was considering either an amazon / ebay "special" logic analyser, or trying this on a spare board: https://github.com/aster94/logic-analyzer.  
I did try the pass through mode for a bit (can't quite remember how long, but a few hours) and I don't recall seeing it fail.
I was considering putting in a power switch to the display, so if it goes beepy, I can reset it without having to reset at the socket / rcd, or maybe find the buzzer and remove it.  Even with it throwing an E13, it seems that it can still control it

**Comment by cobaltfish** [comment]
Agh Fail!  It rebooted and then while it doing so, caused an E13:

Last boot: 29/04/2022, 17:59:29 Hardware Watchdog

@visualapproach Unfortunately, I wasn't logging the websockets output at the time from my always on linux box, so haven't got any details of how long a pause there was etc. Is it worth me recompiling with all the debug sections uncommented? I've got the serial debug turned on

**Reply by visualapproach** [reply]
Was this with the pass through code or regular? I guess regular since you mentioned web sockets. And we're you using async or regular?

**Reply by cobaltfish** [reply]
sorry, yes, it was regular code, from dev branch, but trying to use the async libs.  It had run quite well until then!

**Comment by cobaltfish** [comment]
@visualapproach another quick question.  I noticed how it seemed to be heating slower on "smart" mode.  When reviewing the logs, I spotted some differences:

|setting|fromdisp|fromcpl
|-------|---------|-------|
|standby|55 1 0 30 0 31 aa|55 1 0 30 0 31 aa|
|pump on|55 1 85 30 0 b6 aa|55 1 85 30 0 b6 aa|
|heat on (initial)|55 1 b5 30 0 e6 aa | 55 1 b5 30 0 e6 aa|
| heat on (later)| 55 1 f5 30 0 26 aa | |

Do I need to do anything other than use the 2 stage heating setup
const uint8_t HEATBITMASK1 =   B00110000;  //48  heater stage 1 = 50%
const uint8_t HEATBITMASK2 =   B01000000;  //64  heater stage 2 = 100%

**Reply by visualapproach** [reply]
Don't think so, but what model did you define? And by cpl you mean the esp?

**Comment by dasfliege** [comment]
Has anyone ever managed to run this thing "stable" on a 4-wire PalmSprings or Maldives? Seems that all have the same issue with e13?

**Reply by cobaltfish** [reply]
I have a Palm Springs Hydrojet (54144) which I'm pretty happy with now.
I think my biggest bit of advice is to get the leads between the board and the plugs as short as possible. Mine are about 1cm or less rather than the 8-10cm they came as. Not the easiest to work with, but seems more reliable.  I've also tried putting the plug for CIO on the "underside" with DSP on the top, as I felt that when inside the unit, it was stretching things a little.
I've tried putting capacitor between gnd and 5v, as well as a 10k Resistor, but at the moment I'm not using that board, and it doesn't seem to have made things worse.
I got a couple of the ferrite cores on the wires on the pump and display, but I suspect they may just be giving a placebo effect.

I'm currently running from the dev v4 branch, and while I did have quite a few issues to begin with, it turned out to be a bad soldering job on one joint to / from the display causing intermittent success.

**Reply by dasfliege** [reply]
Thanks, good to know. I will try shortening the cables. I'm pretty good at soldering, as i do it almost daily, so i don't think that bad soldering is my problem :-) Do you may have a picture which cable you used for RX/TX (yellow/green)?

**Reply by dasfliege** [reply]
Ohhh maaaan. I've shortened the wires to like 2cm and there we go. No e13 anymore :-)
But... I do get the actual temperature displayed in the UI now and i am still able to control the pump through the display. But nothing happens when i switch on something or change temperature though the UI. What could be the problem here? I guess RX/TX should be in the right order, as everything is still working perfect when i use the buttons on the display?

![image](https://user-images.githubusercontent.com/31273000/231706413-618db22c-ec6b-4716-8804-6a6933a8dbaa.png)

**Reply by cobaltfish** [reply]
Have you tried switching on the "Take Control" button?
The board is basically sitting in the middle, and when take control (also in the code as "god mode") is off, it forwards the messages between the pump and display, and button presses work.
When "take control" is on, it filters the command button presses on the display, and gives its own ones, giving you remote control (but still makes sure the display gets the current temp).

**Reply by dasfliege** [reply]
Yeah i did. But it switches back to "off" after a few seconds. Same happens for f.e. the bubbles button, if i activate it while having "take control" active, it just returns to "off" without the pump doing anything.

Also the the Titlebar "Lay-Z-Spa Module" turns to a red background. Does that has any meaning?
EDIT: Seems this is just indicating that websockets has been temporaily disconnected?
![image](https://user-images.githubusercontent.com/31273000/231745440-207223a2-e36e-4d74-895e-2356c30e4fed.png)


![image](https://user-images.githubusercontent.com/31273000/231712923-8825bb4c-cfbf-486d-b062-4c0818d29272.png)

**Reply by dasfliege** [reply]
Just seen that there is a new beta available. Updated it and control in the UI are working now!
Just one more question: I've noticed that set temperatur stil isn't synchronized between the display and the UI. Should that be the case? I now have set target temprature in the display to 30 and in the UI to 25. So which temperature will be the one that matters?

**Reply by visualapproach** [reply]
Take control ON - UI
Take control OFF - display

**Reply by dasfliege** [reply]
Allright. So if take control is "off", the UI/ ESP does not has any funtionality or impact except of displaying actual temperature and submitting MQTT. Correct?

**Reply by visualapproach** [reply]
I guess you could say so. The ESP still does a lot of things like calculating time to ready etc, but it doesn't control the pump.

**Reply by dasfliege** [reply]
Thanks a lot! Here's a picture of my cabling, in case someone else has the same problems. 
Mine is running now without any error since more then 24hours. I did not used any capacitors or resistors at all:

![image](https://user-images.githubusercontent.com/31273000/232024629-63ec75ea-50ad-4be8-b7f5-83af1f121b46.png)

**Reply by Xamude** [reply]
I have run  
Firmware: 2023-04-24-1854
Model: NO54173
V1 Board
for 5 Days and still without E13, last summer never get over 10 hours.
Pump is not connected to the SPA at the moment, so i can´t test everything.
Next i try it with a worse Wifi connection to meet more the real conditions outside.

**Reply by pH03NIxM** [reply]
> Thanks a lot! Here's a picture of my cabling, in case someone else has the same problems. 

Thanks! I have same 54144 pump (hydrojet only, no airjet, and with salt-chlorinator which is working fine since 2018) and v2 board without any resistors/caps and short cables is fine. Thanks for the idea.
